source/slang/slang-check-constraint.cpp

```diff
@@ -1244,58 +1423,152 @@ bool SemanticsVisitor::TryUnifyTypes(
         return TryUnifyConjunctionType(constraints, unifyCtx, fst, snd);
     }
 
-    // If one of the types is a type pack, we need to recursively unify the element types.
-    if (auto fstTypePack = as<ConcreteTypePack>(fst))
+    // Unwrap ConcreteTypePack and unify types with regard to ExpandType.
+    //
+    // Consider this case:
+    // left = ConcreteTypePack(ExpandType, ExpandType)
+    // right = ConcreteTypePack(int, bool, float, double).
+    //
+    // In this case, we shouldn't be mapping the first ExpandType to int and the second
+    // ExpandType to bool, float, double. Instead, they should evenly divide the second type
+    // pack, so we map first ExpandType with int, bool, and second ExpandType to float, double.
+    //
+    if ((as<ConcreteTypePack>(fst) && as<ConcreteTypePack>(snd)) ||
+        (as<ConcreteTypePack>(fst) && as<ExpandType>(snd)) ||
+        (as<ExpandType>(fst) && as<ConcreteTypePack>(snd)))
     {
-        if (auto sndTypePack = as<ConcreteTypePack>(snd))
+        // Generate mapping between the two type lists
+        ShortList<IndexSpanPair> typeMapping;
+        ShortList<Type*> flattenedFirst;
+        ShortList<Type*> flattenedSecond;
+        if (!matchTypeArgMapping(fst, snd, flattenedFirst, flattenedSecond, typeMapping))
+            return false;
+
+        // Apply unification based on the mapping
+        for (const auto& mapping : typeMapping)
         {
-            if (fstTypePack->getTypeCount() != sndTypePack->getTypeCount())
-                return false;
-            for (Index i = 0; i < fstTypePack->getTypeCount(); ++i)
+            SLANG_ASSERT(mapping.first.count > 0 && mapping.second.count > 0);
+            SLANG_ASSERT(mapping.first.count == 1 || mapping.second.count == 1);
+
+            // There will be three possible cases: 1:1, 1:N and N:1.
+            // Determine which side is single and which is multiple
+            bool firstIsSingle = (mapping.first.count == 1);
+
+            // Get the single type and the range for multiple types
+            Type* singleType;
+            Index multipleStartIndex;
+            Index multipleCount;
+            ShortList<Type*>* multipleSideList;
+            bool singleSideIsLeftValue;
+            bool multipleSideIsLeftValue;
+
+            if (firstIsSingle)
```

Comment:
Now that we already have the mapping, why isn't the logic here just as simple as:

1. leftType = firstIsSingle?flattenedFirst[mapping.first.index]:createTypePack(mapping.first);
2. rightType = secondIsSingle?flattenedFirst[mapping.second.index]:createTypePack(mapping.second);
3. TryUnifyTypes(leftType, rightType)

Reply:
There are details about propagating isLValue, but you can just grab that from the `flattenedFirst[mapping.first.index]` in both single and multiple cases.